### PR TITLE
Fix decoration of private services

### PR DESCRIPTION
**Pull request by blazarecki**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | no |
| License | MIT |
| Doc PR | no |

This PR allow to decorate private service.
The alternative is to not remove private aliases which are not use elsewhere.

What do you think ?

### CODE REVIEW

**Comment by stof** [comment]
Please describe the issue you are facing. This change looks wrong to me. When decorating a private service, the service defined for this id must stay private.

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/IntegrationTest.php [hunk 1]
```diff
@@ -113,4 +114,24 @@ public function testProcessInlinesWhenThereAreMultipleReferencesButFromTheSameDe
...
+        $foo = $container->register('foo', '\stdClass');
+        $foo->setPublic(false);
+
+        $bar = $container->register('bar', '\stdClaqs');
+        $bar->setDecoratedService('foo');
+
+        $container->compile();
+
+        try {
+            $container->getAlias('foo');
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/IntegrationTest.php]: getting aliases after compilation does not make sense. ``getAlias`` is not a runtime usage of the container (it is not part of the ContainerInterface).
So this test does not cover a supported usage of the component.

**Comment by dunglas** [comment]
@stof currently it's not possible to decorate a private factory service:

```yaml
services:
services:
    foo_factory:
        class: 'AppBundle\FooFactory'
        public: false

    foo_factory_decorator:
        class: 'AppBundle/FooFactoryDecorator'
        public: false
        decorates: 'foo_factory'

    foo:
        class:  'AppBundle\Foo'
        factory: [ '@foo_factory', 'createFoo' ]
```

The decorator pass create a private alias and it is never made public, and is then removed. This "fix" is naive but works, however it probably create some side effects.
A better fix would be to resolve the alias.

**Comment by dunglas** [comment]
We should probably be checked if an alias is used as a factory in `RemoveUnusedDefinitionsPass`.

**Comment by stof** [comment]
@dunglas there is currently a bug in stable versions with using private aliases as a factory service. This is totally unrelated to the usage of decoration (it happens with any private alias) and has already been fixed in all maintained branches 1 week ago (the issue has been detected by a change in FOSUserBundle).

**Comment by dunglas** [comment]
@stof Thanks this is exactly the fix we need. (The related PR is #17554).

**Comment by stof** [comment]
OK, so I'm closing this PR as the actual issue is already fixed in dev versions

**Pull request closed without merging**